### PR TITLE
4 thread parallelism by default

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
The Session default here is super unusual but seems to both compile and do what we expect as best as I can tell.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc https://internals.rust-lang.org/t/plan-to-test-parallel-rustc/11487

**Comment by alexcrichton** [comment]
Heh looks like one of our tests may be a little different in parallel mode!

In any case r=me for tomorrow with tests passing

File: src/librustc_session/config.rs [hunk 1]
```diff
@@ -1362,7 +1362,7 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
     // a sequential compiler for now. This'll likely be adjusted
```
Comment (ehuss) [inline src/librustc_session/config.rs]: Maybe update the comment since it still mentions the default is 1?

**Comment by Mark-Simulacrum** [comment]
I finally understand why I've consistently seen that test fail for the last couple beta backports locally and just not understood how that can possibly be the case.

Hopefully patched the test but I'll check again before I go to sleep... let's hope it deterministically fails, at least -- if not, I'll just delete it for now.

**Comment by Mark-Simulacrum** [comment]
Okay, CI is passing here. It is past the UTC deadline for nightlies, and is ~21 hours till the next nightly is scheduled to be 'forked' off. So, @bors r=alexcrichton

This hopefully ensures that we can get it in before the next nightly drops (and while we're at it, get some testing inside Rust's CI as well).

**Comment by Mark-Simulacrum** [comment]
And since we want this urgently @bors p=100

**Comment by mati865** [comment]
Chocolatey networking issues...
Can somebody retry it?

**Comment by Mark-Simulacrum** [comment]
@bors r=alexcrichton

I suspect this won't actually land (due to clippy, RLS, etc. all having bare `AtomicU64`). It might be that we will want to disable the `dist-mips-linux` builder for tonight's nightly...

File: src/librustc_data_structures/sync.rs [hunk 2]
```diff
@@ -317,7 +317,9 @@ cfg_if! {
         pub use parking_lot::MutexGuard as LockGuard;
         pub use parking_lot::MappedMutexGuard as MappedLockGuard;
 
-        pub use std::sync::atomic::{AtomicBool, AtomicUsize, AtomicU32, AtomicU64};
+        pub use std::sync::atomic::{AtomicBool, AtomicUsize, AtomicU32;
```
Comment (mati865) [inline src/librustc_data_structures/sync.rs]:
```suggestion
        pub use std::sync::atomic::{AtomicBool, AtomicUsize, AtomicU32};
```

**Comment by Mark-Simulacrum** [comment]
Interesting, that looks like a bug in bors -- it said I approved 20cc24b but is testing c0dbd9978fea, a different commit. In any case, it's doing what I wanted.

**Comment by Mark-Simulacrum** [comment]
Hm, so this failure might be spurious (i.e., timing dependent). @alexcrichton -- perhaps you've seen this before?

Specifically, it looks like [this test](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/build.rs#L3003) expects the build script to build first, but (maybe) with a parallel compiler we're not seeing that happen. I suspect that might be because we're capturing all the jobserver tokens pretty quickly so the build script rustc might not be able to start more than one codegen thread (i.e., the main thread) and as such is waiting on the bar crate to finish.

I've kicked off a build locally of this branch with src/tools/cargo as the test target to hopefully dig in a bit deeper and am going to cede CI time to another PR for now.

**Comment by Mark-Simulacrum** [comment]
@bors retry

If the other PR (beta promotion) fails let's make sure this gets testing right away since it might just pass.

**Comment by alexcrichton** [comment]
Yes that looks like a racy test in Cargo, we likely need to fix that

**Comment by ehuss** [comment]
I'm a bit surprised, because it uses `--jobs 1` to ensure the steps are done serially, and cargo should be sorting the dependencies so that it is deterministic.  

Is it possible that the job server is getting confused with `--jobs 1`?  Is `rustc` maybe releasing too many tokens (similar to the `cc` issue described in https://github.com/rust-lang/cargo/issues/7689)?

**Comment by Mark-Simulacrum** [comment]
Yes, it's a known bug (that we should probably file an issue for) that rustc in parallel mode currently releases its implicit token; we have a fix planned but haven't yet gotten around to implementing it yet.

In discussion with @alexcrichton we concluded that it probably makes sense to just disable Cargo's tests since those are likely inherently racy with this bug for now (we can revert this once we get a nightly out).

@bors r=alexcrichton

**Pull request merged**